### PR TITLE
Integrate SSR benchmarks from material ui

### LINKED ISSUE #2118: Integrate SSR benchmarks from material-ui
These ones: https://github.com/mui-org/material-ui/tree/master/packages/material-ui-benchmark

We should add them to our benchmark suite!

### PR DESCRIPTION
**Pull request by mxstbr**
This totally works, just run `yarn run benchmarks` from the root of the repo and you should run both the client-side as well as the server-side benchmarks!

```sh
benchmarks
├── run.js
├── client
│   ├── index.html
│   ├── run-headless.js
│   ├── src/ # Lots of stuff here
│   ├── webpack.config.js
│   └── yarn.lock
└── server
    ├── package.json
    ├── src
    │   └── benchmark.js
    └── yarn.lock
```

Closes #2118

### CODE REVIEW

**Comment by mxstbr** [comment]
Here's the result of a run just now:

```JS
yarn run v1.7.0
$ (cd ../../ && NODE_ENV=production BABEL_ENV=cjs babel-node benchmarks/server/src/benchmark.js)

Starting benchmarks...
Starting benchmark "StyledComponentsButton"...
Starting benchmark "StyledComponentsButton"...
StyledComponentsButton x 12,894 ops/sec ±6.68% (141 runs sampled)
Starting benchmark "EmotionButton"...
EmotionButton x 61,438 ops/sec ±2.53% (149 runs sampled)
Starting benchmark "JssButton cache requests"...
JssButton cache requests x 60,092 ops/sec ±1.90% (149 runs sampled)
Starting benchmark "JssButton cache instances"...
JssButton cache instances x 61,107 ops/sec ±4.44% (159 runs sampled)
Starting benchmark "JssButton no cache"...
JssButton no cache x 16,393 ops/sec ±6.24% (162 runs sampled)
Starting benchmark "JssButton no styles"...
JssButton no styles x 61,788 ops/sec ±3.12% (174 runs sampled)
Starting benchmark "ButtonBase cache instances"...
ButtonBase cache instances x 18,436 ops/sec ±7.15% (139 runs sampled)
Starting benchmark "ButtonBase cache requests"...
ButtonBase cache requests x 17,237 ops/sec ±7.65% (158 runs sampled)
Starting benchmark "ButtonBase no cache"...
ButtonBase no cache x 4,990 ops/sec ±3.39% (153 runs sampled)
Starting benchmark "HocButton"...
HocButton x 76,931 ops/sec ±4.42% (150 runs sampled)
Starting benchmark "NakedButton"...
NakedButton x 102,395 ops/sec ±1.03% (133 runs sampled)
Starting benchmark "ButtonBase cache"...
ButtonBase cache x 17,999 ops/sec ±5.77% (131 runs sampled)
Starting benchmark "ButtonBase ripple"...
ButtonBase ripple x 23,805 ops/sec ±6.96% (145 runs sampled)
Starting benchmark "ButtonBase disableRipple"...
ButtonBase disableRipple x 27,123 ops/sec ±7.72% (155 runs sampled)
Done in 218.32s.
```

Seems like we got some work to do?

**Comment by mxstbr** [comment]
Welp. If I only create the `ServerStyleSheet` once for all the runs we jump up to 117,264 ops/sec 😅 I guess that means `new ServerStyleSheet()` is the bottleneck for SSR perf and we should look at it?

**Comment by probablyup** [comment]
Probably all the cloning

**Comment by mxstbr** [comment]
Yeah that must be it—can we remove that? Why are we even cloning on the server, shouldn't every request be independent? Why would we ever clone on the server?

**Comment by probablyup** [comment]
Good question for @kitten - I'm not super familiar with the cloning stuff. It seemed like it didn't need to exist to me, but he wrote most of it so I'd defer to his judgement.

**Comment by mxstbr** [comment]
@kitten save uuus, why do we need to clone on the server?

I'm guessing it was because we needed to preserve the `injectGlobal` global styles (since that's only ever called once, not every render) but now that all the styles are in the component context it seems like creating a server style sheet should be a relatively cheap operation?!

**Comment by probablyup** [comment]
@mxstbr what if instead we used the existing example app in the benchmark and just used the `ReactDOM` static rendering APIs instead of client side render as a separate "variant".

**Comment by mxstbr** [comment]
To render a suite you have to click a button, which you can't do on the server right?

**Comment by probablyup** [comment]
We'd just render in memory instead of to the DOM

File: benchmarks/server/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@material-ui/benchmark",
```
Comment (oliviertassinari) [inline benchmarks/server/package.json]: Wrong wording

**Comment by probablyup** [comment]
@mxstbr do we want to keep this?

**Comment by probablyup** [comment]
Closing due to lack of interest, happy to reopen later.

**Pull request closed without merging**